libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/AuthenticationUtilities.kt

```diff
@@ -0,0 +1,313 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.auth
+
+import com.salesforce.androidsdk.R.string.sf__generic_authentication_error
+import com.salesforce.androidsdk.R.string.sf__generic_authentication_error_title
+import com.salesforce.androidsdk.R.string.sf__managed_app_error
+import com.salesforce.androidsdk.accounts.UserAccount
+import com.salesforce.androidsdk.accounts.UserAccountBuilder
+import com.salesforce.androidsdk.accounts.UserAccountManager
+import com.salesforce.androidsdk.analytics.EventBuilderHelper.createAndStoreEventSync
+import com.salesforce.androidsdk.app.Features.FEATURE_BIOMETRIC_AUTH
+import com.salesforce.androidsdk.app.Features.FEATURE_SCREEN_LOCK
+import com.salesforce.androidsdk.app.SalesforceSDKManager
+import com.salesforce.androidsdk.auth.OAuth2.IdServiceResponse
+import com.salesforce.androidsdk.auth.OAuth2.TokenEndpointResponse
+import com.salesforce.androidsdk.auth.OAuth2.addAuthorizationHeader
+import com.salesforce.androidsdk.auth.OAuth2.revokeRefreshToken
+import com.salesforce.androidsdk.config.RuntimeConfig.getRuntimeConfig
+import com.salesforce.androidsdk.push.PushMessaging.register
+import com.salesforce.androidsdk.rest.RestClient.clearCaches
+import com.salesforce.androidsdk.security.BiometricAuthenticationManager
+import com.salesforce.androidsdk.security.BiometricAuthenticationManager.Companion.isBiometricAuthenticationEnabled
+import com.salesforce.androidsdk.security.ScreenLockManager
+import com.salesforce.androidsdk.ui.OAuthWebviewHelper.Companion.MUST_BE_MANAGED_APP_PERM
+import com.salesforce.androidsdk.util.SalesforceSDKLogger.e
+import com.salesforce.androidsdk.util.SalesforceSDKLogger.w
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.launch
+import okhttp3.HttpUrl
+import okhttp3.Interceptor
+import okhttp3.Request.Builder
+import org.json.JSONArray
+import org.json.JSONObject
+import java.net.URI
+import java.util.function.Consumer
+
+private const val TAG = "AuthenticationUtilities"
+
+/**
+ * Called when any (User Agent flow, Web Server after PKCE, Native Login, IDP, ect) authentication
+ * method has received a [TokenEndpointResponse] from the server.
+ *
+ * This function:
+ *  * Blocks integration and managed app users.
+ *  * Retrieves the user's identity.
+ *  * Creates an Account.
+ *  * Checks for any CA/ECA settings such as Screen Lock or Biometric Authentication.
+ */
+internal fun onAuthFlowComplete(
```

Comment:
This is basically that clean "Auth v2" interface I have been wanting.  

I left NativeLogin and IDP alone for now but we should migrate them from `OAuthWebviewHelper` to this utility so the former can be completely removed.  All necessary changes would be on internal code so no rush to do it in 13.0.